src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ReflectMethods.java

```diff
@@ -354,6 +355,34 @@ Name lambdaName() {
         return names.fromString("lambda").append('$', names.fromString(String.valueOf(lambdaCount++)));
     }
 
+    Name methodName(JCMethodDecl method) {
+        CharArrayWriter sig = new CharArrayWriter();
+        var sigGen = new Types.SignatureGenerator(types) {
+            @Override
+            protected void append(char ch) {
+                sig.append(ch);
+            }
+            @Override
+            protected void append(byte[] ba) {
+                sig.append(new String(ba));
+            }
+            @Override
+            protected void append(Name name) {
+                sig.append(name.toString());
+            }
+        };
+        for (var p : method.params) {
```

Comment:
I'd suggest to call the signature generator on the type obtained from `method.sym.externalType(types)`, as that will take into account any extra synthetic parameter that might be added later by javac (which will end up in the bytecode) - e.g. for inner class constructors.

Reply:
Could we encode an instance of `MethodRef` in the field name like we do the invoke operation's descriptor? (I believe there are less restrictions in the names of synthetic fields than ones that can be denoted in source. That might be easier to manage on the compile time and run time sides.

Reply:
Technically to create fields  <method name>+<signature>+$op still looks like a quick workaround.
Can we introduce a custom method attribute or annotation?

Reply:
> Could we encode an instance of `MethodRef` in the field name...

Maybe I don't know all the javac tricks, but I understand the `MethodRef` instance is composed a bit later in the process. Here we try to reconstruct the composition of it (or at least the method parameters).

Reply:
Yes, it's a temporary solution We need to think more carefully about a better location such as a method attribute or synthetic annotation, but I think we should do that separately.

Reply:
Honestly, I would go for a simple fix which addresses this annoying problem now, and perhaps discuss on other ways to address this issue in a separate PR?

Reply:
> Could we encode an instance of `MethodRef` in the field name like we do with the invoke operation's descriptor? (I believe there are less restrictions in the names of synthetic fields than ones that can be denoted in source.) That might be easier to manage on the compile time and run time sides.

Are you asking if, instead of serializing the signature, we couldn't just (a) obtain the `MethodRef` object corresponding to the reflectable method and (b) add that to the synthetic field name? Seems possible, and probably provides a better way to pick that one up reflectively.

Reply:
> Seems possible, and probably provides a better way to pick that one up reflectively.

Note that we might still have mismatches between the "source" signature of the method and the "classfile" one because of synthetic parameters. But maybe we can get there when we get there.

Reply:
Yes, it was more in the spirit of a quicker fix with less mangling code.

Reply:
> > Could we encode an instance of `MethodRef` in the field name...
> 
> Maybe I don't know all the javac tricks, but I understand the `MethodRef` instance is composed a bit later in the process. Here we try to reconstruct the composition of it (or at least the method parameters).

I think @PaulSandoz is referring to the code model's `MethodRef` class. You can easily build one in `ReflectMethods` by calling this: `symbolToErasedMethodRef(sym)` where `sym` is the symbol of the reflectable method.

I believe the "bet" is that the `toString` representation of code model's `MethodRef` is good enough to disambiguate and, at the same time, provides an easy way for the reflective code to reconstruct and pick that back up.

Reply:
Implemented as suggested, thank you.